akka-distributed-data/src/main/scala/akka/cluster/ddata/GSet.scala

```diff
@@ -53,15 +56,32 @@ final case class GSet[A](elements: Set[A]) extends ReplicatedData with Replicate
   /**
    * Adds an element to the set
    */
-  def add(element: A): GSet[A] = assignAncestor(copy(elements + element))
+  def add(element: A): GSet[A] = {
+    val newDelta = _delta match {
+      case Some(e) ⇒ Some(new GSet(e.elements + element)(None))
+      case None    ⇒ Some(new GSet[A](Set.apply[A](element))(None))
+    }
+    assignAncestor(new GSet[A](elements + element)(newDelta))
+  }
 
   override def merge(that: GSet[A]): GSet[A] =
     if ((this eq that) || that.isAncestorOf(this)) this.clearAncestor()
     else if (this.isAncestorOf(that)) that.clearAncestor()
     else {
       clearAncestor()
-      copy(elements union that.elements)
+      new GSet[A](elements union that.elements)(None)
```

Comment:
This is OK. Deltas for `GSet` are homoiconic, so if one of the merged sets is a delta we either get a new normal set with all elements from delta merged with it (right-delta merge) or a new delta containing all elements from summary set (left-delta merge). As a full set contains all its deltas this will not lose any deltas.